File: src/cs/AdventOfCode2024/Day1.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AdventOfCode2024;

public class Day1 : IDay
{
    class Response
    {
        [JsonPropertyName("total_distance")]
        public int TotalDistance { get; set; }

        [JsonPropertyName("total_similarity")]
        public int TotalSimilarity { get; set; }
    }

    public string Execute(string input)
    {
        var lines = input.Split("\n");

        List<int> left = new List<int>();
        List<int> right = new List<int>();

        foreach (var line in lines)
        {
            var segments = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
            left.Add(int.Parse(segments[0]));
            right.Add(int.Parse(segments[1]));
        }

        left.Sort();
        right.Sort();

        Response response = new Response();
        response.TotalDistance = 0;

        for(int i = 0; i < left.Count; i++)
        {
            response.TotalDistance += Math.Abs(left[i] - right[i]);
        }

        response.TotalSimilarity = 0;

        foreach (var l in left)
        {
            var leftInRight = 0;

            foreach (var r in right)
            {
                if(l == r)
                {
                    leftInRight++;
                }
            }

            response.TotalSimilarity += l * leftInRight;
        }

        return JsonSerializer.Serialize(response);
    }
}

File: src/py/days/day1.py
import json

from days.day import Day


class Day1(Day):

    def process_input(self, input_data: str) -> str:

        lines: list[str] = input_data.splitlines()
        left_list: list[int] = []
        right_list: list[int] = []

        for line in lines:
            numbers: list[str] = line.split()
            if len(numbers) > 0:
                left_list.append(int(numbers[0]))
                right_list.append(int(numbers[1]))

        right_list.sort()
        left_list.sort()

        total_distance: int = 0
        total_similarity: float = 0

        for i in range(len(left_list)):
            total_distance += abs(left_list[i] - right_list[i])

        for left in left_list:
            found: int = 0
            for right in right_list:
                if left == right:
                    found += 1
            total_similarity += left * found

        data = {
            "part_one": total_distance,
            "part_two": total_similarity
        }

        return json.dumps(data)